### ISSUE TITLE
On Page creation with assigned User no mail is send

### ISSUE THREAD

**Issue by PeterTheOne** [SemanticMediaWiki/SemanticTasks#15]
### Setup and configuration

- SMW version: 3.0.2
- ST version: 9e5b4b310c88533422dd157981a9dab80fe7b664
- MW version: 1.32.1
- PHP version: 7.2.16-1+0~20190307202415.17+stretch~1.gbpa7be82+wmf1 (apache2handler)
- DB (MySQL etc. and version): MariaDB 10.1.38-MariaDB-0+deb9u1

### Issue

Steps to reproduce: Create a page with content: `[[Assigned to::User:Admin]]`.
Expected result: An Email is send.
What happens: No Email is send.

It works on page edit but not on new page.

[`getNewAssignees`](https://github.com/SemanticMediaWiki/SemanticTasks/blob/87006a58d6ed5a98ed3ff2135f2b7a3bd6e2d0dc/src/Assignees.php#L66) is called but `getCurrentAssignees` doesn't return any assignees. Could it be that on PageContentSaveComplete Hook there are no properties saved yet for new pages? Is there a better hook to use instead? I don't know how to explain that.

**Comment by mwjames** [comment]
> is called but `getCurrentAssignees` doesn't return any assignees. Could it
> be that on PageContentSaveComplete Hook there are no properties saved yet
> for new pages? Is there a better hook to use instead? I don't know how to
> explain that.

Without having looked at it, here are a couple of notes:

When creating a new page (or for that matter alter an existing one) we
hold the `SemanticData` (the object that collects all data from the
processing) in an in-memory stack of the ParserOutput and only after
the LinksUpdate hook is run as part of a deferred update is the
actually `Store` update occurring based on in-memory `SemanticData`
instance. The LinksUpdate event was selected to ensure that we got all
data from all possible manipulators that may alter the `SemanticData`
instance. So, yes, it can very well be that `PageContentSaveComplete`
runs to early meaning that LinksUpdate (deferred) hasn't been done yet
and trying to read from the `Store` will lead to no match given that
the data are not yet written.

I hope this provides some guidance otherwise I really have to look at
it which might be difficult given my time constraint.

PS: There is an example [0] in regards to the
`PageContentSaveComplete` that came up in another discussion and
should provide additional inspiration.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/docs/examples/hook.pagecontentsavecomplete.md

Cheers

**Comment by mwjames** [comment]
> PS: There is an example [0] in regards to the `PageContentSaveComplete` that came up in another discussion and should provide additional inspiration.

@PeterTheOne Do you think above comments help you overcome the issue or do you need additional information or suggestions?

**Comment by PeterTheOne** [comment]
@mwjames Your comments help but I haven't made progress with the time I put into it.. As I understand listening to LinksUpdate Hook won't help here either (also one would need to somehow find out which pages was edited/created with that hook). Does the code you linked show how to access the in-memory stack you mentioned or is that the persistent storage? Also some interfaces like `$mwCollaboratorFactory->newEditInfo(...)` seem to have changed.

**Comment by krabina** [comment]
I did some more testing with
SMW version: 3.1.1
ST version: 2.0.0-alpha (62a5ef2)
MW version: 1.31.5

Exactly the behaviour as described initially. On the first edit, no mail is sent. After that, on the second edit, e-mails are sent.

**Comment by krabina** [comment]
#28 is working fine here. looking forward to a merge and release!

**Comment by krabina** [comment]
Update: working fine with SMW 3.1.1
in SMW 3.0.2, I get the following error

```
[Xdfq@e1c7NHw08qIi0DHpgAAAAA] /Spezial:Mit_Formular_bearbeiten/Aufgabe/Krabina_Bernhard_H%C3%B6hergruppierung/Bef%C3%B6rderung_02_Arbeitsvertrag/Urkunde/Bescheid?Aufgabe%5BAufgabe%5D=Arbeitsvertrag%2FUrkunde%2FBescheid&Aufgabe%5BMa%C3%9Fnahme%5D=Krabina+Bernhard+H%C3%B6hergruppierung%2FBef%C3%B6rderung+02 
Error from line 81 of ../extensions/SemanticTasks/src/SemanticTasksMailer.php: Call to undefined method SMW\MediaWiki\MwCollaboratorFactory::newEditInfo()

Backtrace:

#0 ../extensions/SemanticTasks/src/SemanticTasksMailer.php(72): ST\SemanticTasksMailer::getAssignedUserFromParserOutput(WikiPage, Revision, User)
#1 ../extensions/SemanticTasks/SemanticTasks.php(79): ST\SemanticTasksMailer::mailAssigneesUpdatedTask(ST\Assignees, WikiPage, User, WikitextContent, string, integer, NULL, NULL, integer, Revision)
#2 ../includes/Hooks.php(177): SemanticTasks::{closure}(WikiPage, User, WikitextContent, string, integer, NULL, NULL, integer, Revision, Status, boolean, integer)
#3 ../includes/Hooks.php(205): Hooks::callHook(string, array, array, NULL)
#4 ../includes/page/WikiPage.php(1977): Hooks::run(string, array)
#5 [internal function]: WikiPage->{closure}(Wikimedia\Rdbms\DatabaseMysqli, string)
#6 ../includes/libs/rdbms/database/Database.php(3664): call_user_func_array(Closure, array)
#7 ../includes/deferred/AtomicSectionUpdate.php(35): Wikimedia\Rdbms\Database->doAtomicSection(string, Closure)
#8 ../includes/deferred/DeferredUpdates.php(259): AtomicSectionUpdate->doUpdate()
#9 ../includes/deferred/DeferredUpdates.php(210): DeferredUpdates::runUpdate(AtomicSectionUpdate, Wikimedia\Rdbms\LBFactorySimple, string, integer)
#10 ../includes/deferred/DeferredUpdates.php(127): DeferredUpdates::execute(array, string, integer)
#11 ../includes/MediaWiki.php(606): DeferredUpdates::doUpdates(string, integer)
#12 ../includes/MediaWiki.php(575): MediaWiki::preOutputCommit(RequestContext, Closure)
#13 ../includes/MediaWiki.php(877): MediaWiki->doPreOutputCommit(Closure)
#14 ../includes/MediaWiki.php(524): MediaWiki->main()
#15 ../index.php(42): MediaWiki->run()
#16 {main}

```
